File: 138_Copy_List_with_Random_Pointer.py
# Time Complexity : O(n) [n = number of nodes in the list]
# Space Complexity : O(1) [The new list is returned as output]
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
        
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:return None
        node_map = dict()
        curr = head
        
        def clone(node):
            if not node:return None
            if node in node_map:
                return node_map[node]
            new_node = Node(node.val)
            node_map[node] = new_node
            return new_node
            
        while curr is not None:
            copy_node = clone(curr)
            
            copy_node.next = clone(curr.next)
            copy_node.random = clone(curr.random)
            
            curr = curr.next
            
        return node_map[head]

File: _138_CopyListwithRandomPointer.java
// Time Complexity : o(3(n)) whree nis length of list
// Space Complexity : o(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
// add clone nodes in front of original nodes
// attach random pointer of clone nodes using random pointer of origina nodes
// break clone nodes from original lista nd add them together using dummy pointer
class Solution {
    public Node copyRandomList(Node head) {

        if(head == null) return head;

        // add clone nodes in front of original nodes
        Node temp = head;

        while(temp != null){
            Node curr = new Node(temp.val);

            Node next = temp.next;
            temp.next = curr;
            curr.next = next;

            temp = next;
        }

        // attach random pointer of clone nodes using random pointer of origina nodes
        temp = head;
        while( temp != null){
            if(temp.random != null)
                temp.next.random = temp.random.next;
            temp = temp.next.next;
        }

        // break clone nodes from original lista nd add them together using dummy pointer
        Node dummy = new Node(Integer.MIN_VALUE);
        Node temp1 = dummy;

        temp = head;
        while(temp != null){

            Node dummyNext = temp.next;
            Node nextOriginal = dummyNext.next;

            dummy.next = dummyNext;
            dummyNext.next = null;
            dummy = dummyNext;

            temp.next = nextOriginal;
            temp = nextOriginal;

        }

        return temp1.next;
    }
}
